### PR TITLE
feat: indexing works for hello-world

### PR DESCRIPTION
**Pull request by maximilianwerk**
Be aware of the following change suggestions:

- add `__init__` back to the Executor and make the metaclass less magic. (e.g. requests and metas parsing)
- add fully computed workspace property to executor. That might also be moved into the ZED runtime and be pre-computed. Anyhow, we believe providing it on executor level is the right level of abstraction.
- fixed the `requests` decorator, such that you can do:

```
class MyExecutor(Executor):
    @requests(on='seach')
    def my_func(self, *args, **kwargs):
        pass

my_obj = MyExecutor()
my_obj.my_func() <--- did not work before
```

### CODE REVIEW

File: jina/executors/__init__.py [hunk 1]
```diff
@@ -81,6 +72,16 @@ def __init__(awesomeness = 5):
 
     """
 
+    def __init__(self, metas=None, requests=None, *args, **kwargs):
```
Comment (JoanFM) [inline jina/executors/__init__.py:75]: type hinting

File: jina/executors/indexers/dbms/keyvalue.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    def is_exist(self) -> bool:
+        """
+        Check if the database is exist or not
+
+        :return: true if the absolute index path exists, else false
+        """
+        return os.path.exists(self.index_abspath)
+
+    @cached_property
+    def write_handler(self):
```
Comment (JoanFM) [inline jina/executors/indexers/dbms/keyvalue.py:84]: let's use this refactoring to remove so much magic from this maybe?

File: jina/executors/indexers/dbms/keyvalue.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    @property
+    def size(self) -> int:
+        """
+        The number of vectors or documents indexed.
+
+        :return: size
+        """
+        return self._size
+
+    def __getstate__(self):
```
Comment (JoanFM) [inline jina/executors/indexers/dbms/keyvalue.py:117]: this should be removed as no more dumping is happening.

File: jina/executors/indexers/keyvalue.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                        ('', np.int64),
+                    ],
+                ).tobytes()
+            )
+            self._start += l
+            write_handler.body.write(value)
+            self._size += 1
+
+    @requests(on='delete')
+    def delete(self, keys: Iterable[str], *args, **kwargs) -> None:
```
Comment (JoanFM) [inline jina/executors/indexers/keyvalue.py:176]: this should receive `doc_array`

File: jina/executors/indexers/keyvalue.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+    def get_query_handler(self) -> '_ReadHandler':
+        """
+        Get read file handler.
+
+        :return: read handler.
+        """
+        return _ReadHandler(self.index_abspath, self.key_length)
+
+    def _add(
```
Comment (JoanFM) [inline jina/executors/indexers/keyvalue.py:148]: same here

File: jina/executors/indexers/dbms/keyvalue.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+            size=len(ids),
+            data=self._get_generator(ids),
+        )
+        self.query_handler.close()
+        self.handler_mutex = False
+        # noinspection PyPropertyAccess
+        del self.query_handler
+
+    @requests(on='/index')
+    def add(self, docs: DocumentArray, *args, **kwargs):
```
Comment (JoanFM) [inline jina/executors/indexers/dbms/keyvalue.py:204]: `traversal_path` should be handled

File: jina/executors/__init__.py [hunk 7]
```diff
@@ -81,6 +72,16 @@ def __init__(awesomeness = 5):
...
+        super().__init__()
```
Comment (hanxiao) [inline jina/executors/__init__.py:76]:
```suggestion
```
unnecessary

**Review by JoanFM — CHANGES_REQUESTED** [review]

**Comment by JoanFM** [comment]
Isnt it missing the logic from VectorSearch and VectorIndexDriver?

**Comment by hanxiao** [comment]
im taking this branch fixing what @JoanFM proposed and reverting back this change
> add __init__ back to the Executor and make the metaclass less magic. (e.g. requests and metas parsing)

The use of metaclass can't be fully removed and I don't want to add anymore boilerplate when subclassing an `Executor`.

File: jina/executors/indexers/dbms/keyvalue.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+    def _get_generator(
+        self, ids: List[str]
+    ) -> Generator[Tuple[str, np.array, bytes], None, None]:
+        for id_ in ids:
+            vecs_metas_bytes = super()._query(id_)
+            vec, meta = pickle.loads(vecs_metas_bytes)
+            yield id_, vec, meta
+
+    @requests(on='/dump')
+    def dump(self, parameters, *args, **kwargs) -> None:
```
Comment (hanxiao) [inline jina/executors/indexers/dbms/keyvalue.py:180]: this is wrong, arguments are passed in order. see https://github.com/jina-ai/jina/blob/devel-2.0/COOKBOOK.md#method-arguments

File: jina/resources/fashion/helloworld.indexer.yml [hunk 9]
```diff
@@ -1,16 +1,4 @@
...
-      index_filename: vec.gz
-      metric: euclidean
-    metas:
-      name: vecidx  # a customized name
-  - jtype: KeyValueIndexer
-    with:
-      index_filename: chunk.gz
-    metas:
-      name: chunkidx
+jtype: BinaryPbDBMSIndexer
```
Comment (hanxiao) [inline jina/resources/fashion/helloworld.indexer.yml:1]: this looks unmatch with prev. I don't think the vector indexing is implemented.

**Comment by hanxiao** [comment]
i will merge this PR but I don't think this matches the behavior of indexing at 1.x.

File: jina/executors/__init__.py [hunk 10]
```diff
@@ -81,6 +72,16 @@ def __init__(awesomeness = 5):
...
+        if metas is None:
+            metas = {}
+        if requests is not None:
```
Comment (hanxiao) [inline jina/executors/__init__.py]: wrong logic here

**Pull request merged**